channels/rail/server/rail_main.c

```diff
@@ -0,0 +1,1616 @@
+	                           RAIL_PDU_HEADER_LENGTH, &bytesReturned))
+	{
+		if (GetLastError() == ERROR_NO_DATA)
+			return ERROR_NO_DATA;
+
+		WLog_ERR(TAG,  "channel connection closed");
+		return ERROR_INTERNAL_ERROR;
+	}
+
+	/* Parse header */
+	if ((status = rail_read_pdu_header(s, &orderType, &orderLength)) != CHANNEL_RC_OK)
+	{
+		WLog_ERR(TAG, "rail_read_pdu_header failed with error %"PRIu32"!", status);
+		return status;
+	}
+
+	WLog_DBG(TAG, "Received %s PDU, length:%"PRIu16"",
+	         RAIL_ORDER_TYPE_STRINGS[((orderType & 0xF0) >> 3) + (orderType & 0x0F)], orderLength);
+
+	switch (orderType)
+	{
+		case TS_RAIL_ORDER_HANDSHAKE:
+			{
+				RAIL_HANDSHAKE_ORDER handshake;
+				return rail_recv_client_handshake_order(context, &handshake, s);
+			}
+
+		case TS_RAIL_ORDER_CLIENTSTATUS:
+			{
+				RAIL_CLIENT_STATUS_ORDER clientStatus;
+				return rail_recv_client_client_status_order(context, &clientStatus, s);
+			}
+
+		case TS_RAIL_ORDER_EXEC:
+			{
+				RAIL_EXEC_ORDER exec;
+				return rail_recv_client_exec_order(context, &exec, s);
+			}
+
+		case TS_RAIL_ORDER_SYSPARAM:
+			{
+				RAIL_SYSPARAM_ORDER sysparam;
+				return rail_recv_client_sysparam_order(context, &sysparam, s);
+			}
+
+		case TS_RAIL_ORDER_ACTIVATE:
+			{
+				RAIL_ACTIVATE_ORDER activate;
+				return rail_recv_client_activate_order(context, &activate, s);
+			}
+
+		case TS_RAIL_ORDER_SYSMENU:
+			{
+				RAIL_SYSMENU_ORDER sysmenu;
+				return rail_recv_client_sysmenu_order(context, &sysmenu, s);
+			}
+
+		case TS_RAIL_ORDER_SYSCOMMAND:
+			{
+				RAIL_SYSCOMMAND_ORDER syscommand;
+				return rail_recv_client_syscommand_order(context, &syscommand, s);
+			}
+
+		case TS_RAIL_ORDER_NOTIFY_EVENT:
+			{
+				RAIL_NOTIFY_EVENT_ORDER notifyEvent;
+				return rail_recv_client_notify_event_order(context, &notifyEvent, s);
+			}
+
+		case TS_RAIL_ORDER_WINDOWMOVE:
+			{
+				RAIL_WINDOW_MOVE_ORDER windowMove;
+				return rail_recv_client_window_move_order(context, &windowMove, s);
+			}
+
+		case TS_RAIL_ORDER_SNAP_ARRANGE:
+			{
+				RAIL_SNAP_ARRANGE_ORDER snapArrange;
+				return rail_recv_client_snap_arrange_order(context, &snapArrange, s);
+			}
+
+		case TS_RAIL_ORDER_GET_APPID_REQ:
+			{
+				RAIL_GET_APPID_REQ_ORDER getAppidReq;
+				return rail_recv_client_get_appid_req_order(context, &getAppidReq, s);
+			}
+
+		case TS_RAIL_ORDER_LANGBARINFO:
+			{
+				RAIL_LANGBAR_INFO_ORDER langbarInfo;
+				return rail_recv_client_langbar_info_order(context, &langbarInfo, s);
+			}
+
+		case TS_RAIL_ORDER_LANGUAGEIMEINFO:
+			{
+				RAIL_LANGUAGEIME_INFO_ORDER languageImeInfo;
+				return rail_recv_client_language_ime_info_order(context, &languageImeInfo, s);
+			}
+
+		case TS_RAIL_ORDER_COMPARTMENTINFO:
+			{
+				RAIL_COMPARTMENT_INFO_ORDER compartmentInfo;
+				return rail_recv_client_compartment_info(context, &compartmentInfo, s);
+			}
+
+		case TS_RAIL_ORDER_CLOAK:
+			{
+				RAIL_CLOAK_ORDER cloak;
+				return rail_recv_client_cloak_order(context, &cloak, s);
```

Comment:
recv those types then what ?

Reply:
Again, this is how they do this everywhere in their code. The recv methods call the callbacks internally.